### PR TITLE
Cloud Optimised Point Cloud (COPC) support

### PR DESCRIPTION
**Pull request by Desplandis**
## Description

This PR adds support of [COPC format](https://copc.io/):
- Add [copc.js](https://github.com/connormanning/copc.js) dependency for COPC parsing using [laz-perf](https://github.com/hobuinc/laz-perf) under the hood.
- Expose the following module to users of itowns: `CopcSource` and `CopcLayer` which extends existing `PointCloudLayer`.
- Add an example to load remote COPC file (with some examples from [copc.io](https://copc.io/#example-data)).
- Load the `laz-perf.wasm` by default through a CDN but this could be override by the user.

Note that this PR does not:
- use web worker for parsing (this will be the object of another PR).
- itowns/itowns#2157.
- fix issues in the point cloud engine (SSE computation, point budget, ...).

### Autzen Stadium
image
### SoFI Stadium
image
image

### CODE REVIEW

**Comment by Desplandis** [comment]
This PR is still missing some unit tests but is ready for review.

File: examples/copc_simple_loader.html [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                    view.notifyChange();
+                    layer.delete();
+                }
+
+                layer = new itowns.CopcLayer('COPC', {
+                    source,
+                    crs: view.referenceCrs,
+                    sseThreshold: 2,
+                    pointBudget: 3000000,
+                    // supportsProgressiveDisplay: true,
```
Comment (jogarnier) [inline examples/copc_simple_loader.html]: Commented line to remove ?

File: src/Parser/LAZParser.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    for (let i = 0; i < view.pointCount; i++) {
+        // `getPosition` apply scale and offset transform to the X, Y, Z values.
+        // See https://github.com/connormanning/copc.js/blob/master/src/las/extractor.ts.
+        const [x, y, z] = getPosition.map(f => f(i));
+        positions[i * 3] = x;
+        positions[i * 3 + 1] = y;
+        positions[i * 3 + 2] = z;
+
+        const intensity = getIntensity(i);
+        intensities[i] = intensity;
```
Comment (mgermerie) [inline src/Parser/LAZParser.js]: Maybe you could avoid creating a `const` that is only used on line 34, and replace line 34 with 
```js
intensities[i] = getIntensity(i);
```
I don't know a lot about memory management but do you know if there is a difference in creating constants at each loop iteration ?

Comment (mgermerie) [reply src/Parser/LAZParser.js]: The same question goes for lines 36 to 39.

Comment (mgermerie) [reply src/Parser/LAZParser.js]: And 59 to 63 ;)

**Comment by jogarnier** [comment]
I had several eslint errors while doing `npm start` on my local code, please have a look over the following : 

```
itowns\src\Core\Style.js

  4:1  error  `@mapbox/mapbox-gl-style-spec` import should occur before import of `Core/Feature`  import/order
  5:1  error  `three` import should occur before import of `Core/Feature`                         import/order

itowns\src\Parser\B3dmParser.js
  10:1  error  `three` import should occur before import of `Core/System/Capabilities`  import/order

itowns\utils\debug\Debug.js
  5:1  error  `three` import should occur before import of `Core/Geographic/Coordinates`  import/order


itowns\src\Source\CopcSource.js
  6:1  error  `three` import should occur before import of `Core/Geographic/Extent`  import/order
```

File: src/Source/CopcSource.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
+import { Binary, Info, Las } from 'copc';
+import Extent from 'Core/Geographic/Extent';
+import Fetcher from 'Provider/Fetcher';
+import LAZParser from 'Parser/LAZParser';
+import Source from 'Source/Source';
+import * as THREE from 'three';
```
Comment (jogarnier) [inline src/Source/CopcSource.js:6]: I had several eslint errors while doing `npm start` on my local code, please have a look over the following : 

```
itowns\src\Core\Style.js

  4:1  error  `@mapbox/mapbox-gl-style-spec` import should occur before import of `Core/Feature`  import/order
  5:1  error  `three` import should occur before import of `Core/Feature`                         import/order

itowns\src\Parser\B3dmParser.js
  10:1  error  `three` import should occur before import of `Core/System/Capabilities`  import/order

itowns\utils\debug\Debug.js
  5:1  error  `three` import should occur before import of `Core/Geographic/Coordinates`  import/order


itowns\src\Source\CopcSource.js
  6:1  error  `three` import should occur before import of `Core/Geographic/Extent`  import/order
```

This a bit out of your scope but if you can handle this, I think it would be nice, 
it refers to this commit a8413433091882be7c461977ccdb343c75d89d27, where some npm packages have been updated (eslint / eslint-plugin-import / eslint-webpack-plugin which has taken one major version update)

It's a bit weird that the tests have passed, by the way

File: examples/copc_simple_loader.html [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                });
+                view.addLayer(layer).then(onLayerReady);
+                debug.PotreeDebug.initTools(view, layer, gui);
+            }
+
+            function loadAutzen() {
+                setUrl("https://s3.amazonaws.com/hobu-lidar/autzen-classified.copc.laz");
```
Comment (jogarnier) [inline examples/copc_simple_loader.html:116]: Can I suggest you to upload the files onto https://github.com/iTowns/iTowns2-sample-data, instead ?

**Review by jogarnier — APPROVED** [review]

File: examples/copc_simple_loader.html (hunk 4 above)
Comment (Desplandis) [reply examples/copc_simple_loader.html:116]: It seems we can't HTTP range request on `raw.githubusercontent.com` (due to CORS preflight). It is an [issue](https://github.com/orgs/community/discussions/14214) on github side... This is why I use the official examples of the [copc.io](https://copc.io/#example-data) site.

File: examples/copc_simple_loader.html (hunk 1 above)
Comment (Desplandis) [reply examples/copc_simple_loader.html]: I'll remove it

File: src/Parser/LAZParser.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        classification: classifications,
+        pointSourceID: pointSourceIDs,
+        color: colors,
+    };
+}
+
+function parseBuffer(pointData, options) {
+    const { header, eb } = options.in;
+    const view = copc.Las.View.create(pointData, header, eb);
+    const attrs = parseView(view);
```
Comment (mgermerie) [inline src/Parser/LAZParser.js]: It seems that this is the only call of `parseView` method. It does not pass an `options` parameter. Maybe you could add it so that the `options.colorDepth` specified by the user when instantiating its `CopcSource` is taken into account ?

File: src/Layer/CopcLayer.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        this.whenReady = this.source.whenReady.then((/** @type {CopcSource} */ source) => {
+            const { cube, rootHierarchyPage } = source.info;
+            const { pageOffset, pageLength } = rootHierarchyPage;
+
+            this.root = new CopcNode(0, 0, 0, 0, pageOffset, pageLength, this, -1);
+            this.root.bbox.min.fromArray(cube, 0);
+            this.root.bbox.max.fromArray(cube, 3);
+
+            const minElevationRange = source.header.min[2];
+            const maxElevationRange = source.header.max[2];
```
Comment (mgermerie) [inline src/Layer/CopcLayer.js]: Why storing these in constants ?

File: src/Core/CopcNode.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+     * @param {number} size
+     */
+    async _fetch(offset, size) {
+        return this.layer.source.fetcher(this.layer.source.url, {
+            ...this.layer.source.networkOptions,
+            headers: {
+                ...this.layer.source.networkOptions.headers,
+                range: `bytes=${offset}-${offset + size - 1}`,
+            },
+        });
```
Comment (mgermerie) [inline src/Core/CopcNode.js:59]: It might be too tedious for the benefit but do you think this part could be factorized with the first promise of the `get` method in `CopcSource`

File: src/Core/CopcNode.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            const z = node.z * 2;
+
+            node.findAndCreateChild(depth, x,     y,     z,     hierarchy, stack);
+            node.findAndCreateChild(depth, x + 1, y,     z,     hierarchy, stack);
+            node.findAndCreateChild(depth, x,     y + 1, z,     hierarchy, stack);
+            node.findAndCreateChild(depth, x + 1, y + 1, z,     hierarchy, stack);
+            node.findAndCreateChild(depth, x,     y,     z + 1, hierarchy, stack);
+            node.findAndCreateChild(depth, x + 1, y,     z + 1, hierarchy, stack);
+            node.findAndCreateChild(depth, x,     y + 1, z + 1, hierarchy, stack);
+            node.findAndCreateChild(depth, x + 1, y + 1, z + 1, hierarchy, stack);
```
Comment (mgermerie) [inline src/Core/CopcNode.js:168]: As we discussed, could you return the child in `findAndCreateChild` and push it into `stack` instead of passing `stack` as a parameter ?

File: src/Core/CopcNode.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+        let pointCount;
+        let offset;
+        let byteSize;
+
+        const node = hierarchy.nodes[id];
+        if (node) {
+            pointCount = node.pointCount;
+            offset = node.pointDataOffset;
+            byteSize = node.pointDataLength;
```
Comment (mgermerie) [inline src/Core/CopcNode.js:114]: I might have missed something but I figure that doing this results in all 8 nodes from depth 1 (assuming the root node is at depth 0) will have the same number of points, entry offset and size (with equal `numPoints`, `entryOffset` and `entryLength` properties values). These common values are those of the parent node (root in my example). Is this intended ?

File: src/Core/CopcNode.js (hunk 8 above)
Comment (mgermerie) [reply src/Core/CopcNode.js:168]: This mechanism is also present in `EntwinePointTileNode`.

File: src/Provider/PointCloudProvider.js [hunk 10]
```diff
@@ -36,7 +36,12 @@ export default {
             addPickingAttribute(points);
             points.frustumCulled = false;
             points.matrixAutoUpdate = false;
-            if (!layer.isEntwinePointTileLayer) {
+            if (!layer.isEntwinePointTileLayer && !layer.isCopcLayer) {
```
Comment (mgermerie) [inline src/Provider/PointCloudProvider.js]: If I understand well, this test would be removed should we just add `this.scale = new THREE.Vector3(1.0, 1.0, 1.0)` in `EntwinePointCloudLayer.whenReady` ?

File: src/Layer/CopcLayer.js (hunk 6 above)
Comment (mgermerie) [reply src/Layer/CopcLayer.js]: Also, I could not find an usage of `minElevationRange` and `maxElevationRange` properties, nor some documentation. What are they for ?

File: examples/copc_simple_loader.html [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            let layer; // COPCLayer
+
+            const uri = new URL(location);
+
+            const gui = new dat.GUI();
+
+            const viewerDiv = document.getElementById('viewerDiv');
+            const view = new itowns.View('EPSG:4326', viewerDiv);
+            const controls = new itowns.PlanarControls(view);
+            viewerDiv.style.display = 'block';
```
Comment (mgermerie) [inline examples/copc_simple_loader.html]: Why is this needed ?

**Review by mgermerie — COMMENTED** [review]
Thanks for this PR ! 
I only have a few comments and two general remarks : 
- Could you add yourself to CONTRIBUTORS.md ?
- I looked at coverage reports, and strangely some parts are not covered even if you call them in your new unit tests (the constructor of `CopcLayer` for instance). Could you take a look at it ? and add tests if needed ?

Other than that it should be good for me !

File: src/Layer/CopcLayer.js (hunk 6 above)
Comment (Desplandis) [reply src/Layer/CopcLayer.js]: Leftovers of when this PR also added support of other LAS Attributes. I will remove them for now.

File: src/Parser/LAZParser.js (hunk 2 above)
Comment (Desplandis) [reply src/Parser/LAZParser.js]: I suppose the javascript engine is smart enough to optimize such construct. However, they are leftovers of debugging, I will directly set those values.

File: src/Parser/LAZParser.js (hunk 5 above)
Comment (Desplandis) [reply src/Parser/LAZParser.js]: Whoops... Forget to pass an the `options` parameter.
The `parseView` function is only called once because I may, in another PR, add a `parseFile` function to parse an entire LAS/LAZ file and replace entirely the `LASParser` file.

File: src/Provider/PointCloudProvider.js (hunk 10 above)
Comment (Desplandis) [reply src/Provider/PointCloudProvider.js]: You'll also need to check that `node.bbox.min` shall not create unwanted offsets...

File: src/Source/CopcSource.js (hunk 3 above)
Comment (Desplandis) [reply src/Source/CopcSource.js:6]: This is caused by recent versions of `eslint` and will be fixed in another PR.

File: src/Core/CopcNode.js (hunk 8 above)
Comment (Desplandis) [reply src/Core/CopcNode.js:168]: I agree. Should I modify `EntwinePointTileNode` too ?

File: src/Core/CopcNode.js (hunk 7 above)
Comment (Desplandis) [reply src/Core/CopcNode.js:59]: I used `this.layer.source.fetcher` directly since I supposed that a user could use a custom `fetcher` or `Source`. Maybe this is too far-fetched!

File: examples/copc_simple_loader.html (hunk 11 above)
Comment (Desplandis) [reply examples/copc_simple_loader.html]: Removed it.

File: src/Provider/PointCloudProvider.js (hunk 10 above)
Comment (Desplandis) [reply src/Provider/PointCloudProvider.js]: I opened an issue on this topic. See #2189 !

File: src/Core/CopcNode.js (hunk 9 above)
Comment (Desplandis) [reply src/Core/CopcNode.js:114]: Fixed it!

**Comment by Desplandis** [comment]
@mgermerie @jogarnier Since your last review, I moved part of `LAZParser.js` to  `LAZLoader.js`. This is done to split the parser in two parts: one that could be "workerisable" (i.e. `LAZLoader`) in a future PR (I already have some POC) and one that creates Geometry Buffers/instantiate workers.

I also took into account all your comments.

This is ready for a second (quick) review!

**Comment by Desplandis** [comment]
@mgermerie @jogarnier I also opened an issue on the copc.js library (see connormanning/copc.js#6)

TLDR: `copc.js` lazy-load the fs node module which is not platform agnostic. This causes bundling issues (with at least webpack and vite) in a similar fashion as our issue #2200.

If this issue is not fixed upstream, we may have to "fork" their package to remove the dozen lines causing this problem...

**Comment by connormanning** [comment]
I've left a comment on the referenced ticket, but regardless I don't think you'd need a fork to accomplish this.  You can accomplish this downstream of copc.js with a simple library whose code simply re-exports everything from copc.js, and whose bundler nullfies any reference to `fs`, for example with webpack's `null-loader`, or using a Rollup IIFE bundle that sets the `fs` global to `null` since the `fs` code-paths would not be executed in a web environment anyway.  You could also do this in the itowns config rather than a separate library.

I think the capability to work around this exists - if you find yourself needing to make a fork over this let me know and I can lend a hand.

File: examples/copc_simple_loader.html [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+            <input type="text" id="url" />
+            <button onclick="readURL()">Load</button>
+            <div>
+                <button onClick="loadAutzen()">Load Autzen Stadium (80mb)</button>
+                <button onClick="loadSofi()">Load SoFI Stadium (2.3gb)</button>
+                <button onClick="loadMillsite()">Load Millsite (1.9gb)</button>
+            <div id="share"></div>
+            </div>
+        </div>
+        <div id="viewerDiv">
```
Comment (AnthonyGlt) [inline examples/copc_simple_loader.html]: missing closing div

File: examples/copc_simple_loader.html [hunk 13]
```diff
@@ -0,0 +1,128 @@
+<html>
+    <head>
+        <meta charset="UTF-8">
+
+        <title>Itowns - COPC loader</title>
+
+        <link rel="stylesheet" type="text/css" href="css/example.css">
+        <link rel="stylesheet" type="text/css" href="css/LoadingScreen.css">
+
+        <style type="text/css">
```
Comment (AnthonyGlt) [inline examples/copc_simple_loader.html]: type is deprecated/unnecessary https://developer.mozilla.org/en-US/docs/web/html/element/style#type

File: src/Core/CopcNode.js [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        const f = 2 ** (node.depth - this.depth);
+
+        // size of the child node bbox (Vector3), based on the size of the
+        // parent node, and divided by the factor
+        this.bbox.getSize(size).divideScalar(f);
+
+        // initialize the child node bbox at the location of the parent node bbox
+        node.bbox.min.copy(this.bbox.min);
+
+        // position of the parent node, if it was at the same depth than the
```
Comment (AnthonyGlt) [inline src/Core/CopcNode.js]:
```suggestion
        // position of the parent node, if it was at the same depth as the
```

File: src/Parser/LASParser.js [hunk 15]
```diff
@@ -22,6 +51,39 @@ export default {
...
+
+
+    /**
+     * Parses a chunk of a LAS or LAZ (LASZip) and returns the corresponding
+     * `THREE.BufferGeometry`.
+     *
+     * @param {ArrayBuffer} data - The file content to parse.
+     * @param {Object} options
+     * @param {Object} options.in - Options to give to the parser.
+     * @param {number} option.in.pointCount - Number of points encoded in this
```
Comment (AnthonyGlt) [inline src/Parser/LASParser.js]:
```suggestion
     * @param {number} options.in.pointCount - Number of points encoded in this
```

File: src/Provider/PointCloudProvider.js [hunk 16]
```diff
@@ -36,7 +36,12 @@ export default {
...
+                // TODO: Harmonize properties of all PointCloudLayers
+                // Since both EPT and COPC Layer apply scaling at parsing, they
+                // should set a `layer.scale` at `THREE.Vector(1.0, 1.0, 1.0)`.
+                // For the `offset`, see if possible to move this property
+                // upward in Potree layer.
```
Comment (AnthonyGlt) [inline src/Provider/PointCloudProvider.js]: maybe add the link to the issue https://github.com/iTowns/itowns/issues/2189 , easier to come back to it

File: src/Source/CopcSource.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        [];
+
+    return { header, info, wkt, eb };
+}
+
+/**
+ * @classdesc
+ * A source for [Cloud Optimised Point Cloud](https://copc.io/) (COPC) data.
+ * Such data consists of a [LAZ 1.4](https://www.ogc.org/standard/las/) file
+ * that stores compressed points data organized in a clustered octre.
```
Comment (AnthonyGlt) [inline src/Source/CopcSource.js]:
```suggestion
 * that stores compressed points data organized in a clustered octree.
```

File: src/Source/CopcSource.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+ *
+ * @extends {Source}
+ *
+ * @property {boolean} isCopcSource - Read-only flag to check that a given
+ * object is of type CopcSource.
+ * @property {Object} header - LAS header of the source.
+ * @property {Object[]} eb - List of headers of each Variable Length Records
+ * (VLRs).
+ * @property {Object} info - COPC `info` VLR.
+ * @property {number[]} info.cube - Bouding box of the octree as a 6-elements.
```
Comment (AnthonyGlt) [inline src/Source/CopcSource.js]:
```suggestion
 * @property {number[]} info.cube - Bounding box of the octree as a 6-elements.
```

File: src/Source/CopcSource.js [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+ * root node data chunk.
+ * @property {number} info.rootHierarchyPage.pageOffset - Size (in bytes) of the
+ * root node data chunk.
+ * @property {number[]} gpsTimeRange - A 2-element tuple denoting the minimum
+ * and maximum values of attribute `gpsTime`.
+ */
+class CopcSource extends Source {
+    /**
+     * @param {Object} config - Source configuration
+     * @param {string} config.url - URL of the COPC ressource.
```
Comment (AnthonyGlt) [inline src/Source/CopcSource.js]:
```suggestion
     * @param {string} config.url - URL of the COPC resource.
```
Little mistake made multiple times in the comments

File: src/Source/CopcSource.js [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+                headers: {
+                    ...this.networkOptions.headers,
+                    range: `bytes=${begin}-${end - 1}`,
+                },
+            }).then(buffer => new Uint8Array(buffer));
+        this.whenReady = getHeaders(get).then((metadata) => {
+            this.header = metadata.header;
+            this.info = metadata.info;
+            this.eb = metadata.eb;
+            // TODO: use wkt definition in `metadata.wkt` to infer/define crs
```
Comment (AnthonyGlt) [inline src/Source/CopcSource.js:108]: Is it planned (issue/proposal/PR link) ?

**Comment by LoicUTC** [comment]
That would be a great feature :)

File: src/Provider/PointCloudProvider.js (hunk 16 above)
Comment (Desplandis) [reply src/Provider/PointCloudProvider.js]: Fixed by #2284

File: src/Source/CopcSource.js (hunk 20 above)
Comment (Desplandis) [reply src/Source/CopcSource.js:108]: Not for now, we let the user set the `crs` manually. We would need to have `proj4` expose its `wkt`-parsing ability.

File: examples/copc_simple_loader.html (hunk 12 above)
Comment (Desplandis) [reply examples/copc_simple_loader.html]: Done

**Review by mgermerie — APPROVED** [review]
:tada:
